### PR TITLE
Ensure ordering with retry

### PR DESCRIPTION
**Pull request by dhruvjain99**
## Proposed Changes

Adds test to ensure ordering when fetched from db. Also add retry functionality on failing db ops.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes issue #XXXX)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, styles...)
- [ ] DevOps (Build scripts, pipelines...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [ ] I have read the `CODE_OF_CONDUCT.md` document
- [X] All tests pass locally with my changes
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if needed)
- [ ] Any dependent changes have been merged and published in related repositories
- [ ] I have updated changelog (At the bottom of the release version)
- [ ] I have squashed all my commits into one before merging

### CODE REVIEW

**Review by ajatprabha — COMMENTED** [review]
Should we track retry attempts as metrics?

**Comment by dhruvjain99** [comment]
> Should we track retry attempts as metrics?

@ajatprabha do you want me to add another metric for retry attempts? Though we are already tracking errors and on receiving errors we retry.

**Review by ajatprabha — APPROVED** [review]

File: apps/vmq_server/src/vmq_message_store.erl [hunk 1]
```diff
@@ -22,6 +22,9 @@
          delete/2,
          find/1]).
 
+-define(RETRY_INTERVAL, 2000).
+-define(NR_OF_RETRY, 20).
```
Comment (dumbpreacher) [inline apps/vmq_server/src/vmq_message_store.erl:26]: Isn't 20 a large number of retries? Also, should this be a config?

File: apps/vmq_server/src/vmq_message_store.erl [hunk 2]
```diff
@@ -46,50 +49,74 @@ stop() ->
...
-    ok.
+            vmq_metrics:incr_msg_store_ops_error(write),
+            timer:sleep(?RETRY_INTERVAL),
+            write_with_retry(SubscriberId, Msg, N-1)
+    end;
+write_with_retry(_SubscriberId, _Msg, _N) -> ok.
 
 read(SubscriberId, MsgRef) ->
+    read_with_retry(SubscriberId, MsgRef, ?NR_OF_RETRY).
+read_with_retry(SubscriberId, MsgRef, N) when N > 0 ->
```
Comment (dumbpreacher) [inline apps/vmq_server/src/vmq_message_store.erl:68]: We should be measuring time taken to read for all attempts, not single attempts.

File: apps/vmq_server/src/vmq_message_store.erl [hunk 3]
```diff
@@ -46,50 +49,74 @@ stop() ->
...
+        {error, no_matching_hook_found} = ErrRes -> ErrRes;
+        {error, Err} ->
             lager:error("Error: ~p", [Err]),
             vmq_metrics:incr_msg_store_ops_error(read),
-            E
-    end.
+            timer:sleep(?RETRY_INTERVAL),
+            read_with_retry(SubscriberId, MsgRef, N-1)
+    end;
+read_with_retry(_SubscriberId, _MsgRef, _N) -> {error, retry_exhausted}.
```
Comment (dumbpreacher) [inline apps/vmq_server/src/vmq_message_store.erl:78]: Let's have a metric for this?

**Pull request merged**